File: 2020/10/week3/minhyungPark/algorithm/Leetcode1615.kt
/**
 * Leetcode 1615 - Maximal Network Rank
 * https://leetcode.com/contest/weekly-contest-210/problems/maximal-network-rank/
 */

class Leetcode1615 {
    fun maximalNetworkRank(n: Int, roads: Array<IntArray>): Int {
        
        val connected = Array(n) { BooleanArray(n) }
        val bridgeCount = IntArray(n)
        
        roads.forEach { 
            connected[it[0]][it[1]] = true
            connected[it[1]][it[0]] = true
            bridgeCount[it[0]]++
            bridgeCount[it[1]]++
        }
        
        var result = 0
        
        for (i in 0 until n) {
            for (j in 0 until n) {
                if (i==j) continue
                val tmp = bridgeCount[i] + bridgeCount[j] - if (connected[i][j]) 1 else 0
                result = if (result < tmp) tmp else result
            }
        }
        
        return result
    }
}

File: 2020/10/week4/1ilsang/algorithm/leetcode1615.js
/**
 * 1615. Maximal Network Rank
 * https://leetcode.com/contest/weekly-contest-210/problems/maximal-network-rank/
 * @param {number} n
 * @param {number[][]} roads
 * @return {number}
 */
const maximalNetworkRank = (n, roads) => {
  const count = Array(n).fill(0);

  const mat = Array(n)
    .fill(null)
    .map(() => Array(n).fill(false));

  roads.forEach(([a, b]) => {
    mat[a][b] = true;
    mat[b][a] = true;

    count[a]++;
    count[b]++;
  });

  let res = 0;

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      let aux = count[i] + count[j] - (mat[i][j] ? 1 : 0);
      res = Math.max(aux, res);
    }
  }

  return res;
};